src/pypyjs.js

```diff
@@ -0,0 +1,1112 @@
+            throw new pypyjs.Error('Failed to execute python code');
+          }
+
+          Module._free(code);
+        });
+      });
+    })
+    .then(resolve, reject);
+  });
+}
+
+pypyjs.prototype.inJsModules = null;
+
+// A simple file-fetching wrapper around XMLHttpRequest,
+// that treats paths as relative to the pypyjs.js root url.
+//
+pypyjs.prototype.fetch = function fetch(relpath, responseType) {
+  const rootURL = this.rootURL || pypyjs.rootURL;
+
+  if (this.inJsModules && this.inJsModules[relpath]) {
+    return new Promise((resolve) => {
+      resolve({ responseText: this.inJsModules[relpath] });
+    });
+  }
+
+  // For the web, use XMLHttpRequest.
+  if (typeof XMLHttpRequest !== 'undefined') {
+    return new Promise((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
+      xhr.onload = function onload() {
+        if (xhr.status >= 400) {
+          reject(xhr);
+        } else {
+          resolve(xhr);
+        }
+      };
+
+      xhr.open('GET', rootURL + relpath, true);
+      xhr.responseType = responseType || 'text';
+      xhr.send(null);
+    });
+  }
+
+  // For nodejs, use fs.readFile.
+  if (typeof fs !== 'undefined' && typeof fs.readFile !== 'undefined') {
+    return new Promise((resolve, reject) => {
+      fs.readFile(path.join(rootURL, relpath), (err, data) => {
+        if (err) return reject(err);
+        resolve({ responseText: data.toString() });
+      });
+    });
+  }
+
+  // For spidermonkey, use snarf (which has a binary read mode).
+  if (typeof snarf !== 'undefined') {
+    return new Promise((resolve) => {
+      const data = snarf(rootURL + relpath);
+      resolve({ responseText: data });
+    });
+  }
+
+  // For d8, use read() and readbuffer().
+  if (typeof read !== 'undefined' && typeof readbuffer !== 'undefined') {
+    return new Promise((resolve) => {
+      const data = read(rootURL + relpath);
+      resolve({ responseText: data });
+    });
+  }
+
+  return new Promise((resolve, reject) => {
+    reject('unable to fetch files');
+  });
+};
+
+// add a Module to the vm from a file so you can import it later. Uses a path
+// relative to pypyjs.
+pypyjs.prototype.addModuleFromFile = function addModuleFromFile(name, file) {
+  return this.fetch(file).then((data) => this.addModule(name, data.responseText));
+};
+
+// add a Module to the vm from javascript. It will automatically write it to
+// the "disk" when you import it in your code later.
+pypyjs.prototype.addModule = function addModule(name, source) {
+  return this.findImportedNames(source).then((imports) => {
+    // keep track of any modules that have been previously loaded
+    if (this._loadedModules[name]) {
+      this._modulesToReset[name] = true;
+      this._loadedModules[name] = null;
+    }
+    this._allModules[name] = {
+      file: `${name}.py`,
+      imports
+    };
+    if (!this.inJsModules) {
+      this.inJsModules = {};
+    }
+    this.inJsModules[`modules/${name}.py`] = source;
+  });
```

Comment:
I've been thinking about this `inJsModules` stuff, and I *think* I have my head around it.  Here's a proposed refactor that IMHO works a little more with-the-grain of the existing loading and filesystem machinery.  It may not do exactly what you want though, so interested in your thoughts:

* make `addModule` just write the module data into the appropriate location in the VM's virtual filesystem, either directly using `_writeModuleFile` or some generalization of it.
* make `fetch` look in the virtual filesystem first, only fetching from the server if the file is not found locally

I *think* this would mean that we don't have to maintain an explicit `inJsModules` hash, but can just store state directly in the virtual filesystem.  There may be some subtleties with path-handling though.  Thoughts?